protocols/relay/src/v2/protocol/inbound_hop.rs

```diff
@@ -0,0 +1,237 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use crate::v2::message_proto::{hop_message, HopMessage, Limit, Reservation, Status};
+use crate::v2::protocol::{HOP_PROTOCOL_NAME, MAX_MESSAGE_SIZE};
+use asynchronous_codec::{Framed, FramedParts};
+use bytes::{Bytes, BytesMut};
+use futures::{future::BoxFuture, prelude::*};
+use libp2p_core::{upgrade, Multiaddr, PeerId};
+use libp2p_swarm::NegotiatedSubstream;
+use prost::Message;
+use std::convert::TryInto;
+use std::io::Cursor;
+use std::iter;
+use std::time::{Duration, SystemTime, UNIX_EPOCH};
+use thiserror::Error;
+use unsigned_varint::codec::UviBytes;
+
+pub struct Upgrade {
+    pub reservation_duration: Duration,
+    pub max_circuit_duration: Duration,
+    pub max_circuit_bytes: u64,
+}
+
+impl upgrade::UpgradeInfo for Upgrade {
+    type Info = &'static [u8];
+    type InfoIter = iter::Once<Self::Info>;
+
+    fn protocol_info(&self) -> Self::InfoIter {
+        iter::once(HOP_PROTOCOL_NAME)
+    }
+}
+
+impl upgrade::InboundUpgrade<NegotiatedSubstream> for Upgrade {
+    type Output = Req;
+    type Error = UpgradeError;
+    type Future = BoxFuture<'static, Result<Self::Output, Self::Error>>;
+
+    fn upgrade_inbound(self, substream: NegotiatedSubstream, _: Self::Info) -> Self::Future {
+        let mut codec = UviBytes::<bytes::Bytes>::default();
+        codec.set_max_len(MAX_MESSAGE_SIZE);
+        let mut substream = Framed::new(substream, codec);
+
+        async move {
+            let msg: bytes::BytesMut = substream
+                .next()
+                .await
+                .ok_or_else(|| std::io::Error::new(std::io::ErrorKind::UnexpectedEof, ""))??;
+
+            let HopMessage {
+                r#type,
+                peer,
+                reservation: _,
+                limit: _,
+                status: _,
+            } = HopMessage::decode(Cursor::new(msg))?;
+
+            let r#type = hop_message::Type::from_i32(r#type).ok_or(UpgradeError::ParseTypeField)?;
+            match r#type {
+                hop_message::Type::Reserve => Ok(Req::Reserve(ReservationReq {
+                    substream,
+                    reservation_duration: self.reservation_duration,
+                    max_circuit_duration: self.max_circuit_duration,
+                    max_circuit_bytes: self.max_circuit_bytes,
+                })),
+                hop_message::Type::Connect => {
+                    let dst = PeerId::from_bytes(&peer.ok_or(UpgradeError::MissingPeer)?.id)
+                        .map_err(|_| UpgradeError::ParsePeerId)?;
+                    Ok(Req::Connect(CircuitReq { dst, substream }))
+                }
+                hop_message::Type::Status => Err(UpgradeError::UnexpectedTypeStatus),
+            }
```

Comment:
A question of taste but I like to separate error paths from happy paths:

```suggestion
            let req = match r#type {
                hop_message::Type::Reserve => Req::Reserve(ReservationReq {
                    substream,
                    reservation_duration: self.reservation_duration,
                    max_circuit_duration: self.max_circuit_duration,
                    max_circuit_bytes: self.max_circuit_bytes,
                }),
                hop_message::Type::Connect => {
                    let dst = PeerId::from_bytes(&peer.ok_or(UpgradeError::MissingPeer)?.id)
                        .map_err(|_| UpgradeError::ParsePeerId)?;

                    Req::Connect(CircuitReq { dst, substream })
                }
                hop_message::Type::Status => retrurn Err(UpgradeError::UnexpectedTypeStatus),
            };
            
            Ok(req)
```

Reply:
Suggestion applied via b9eb0f061f939210229eff85eb2e4be7c9ae10a3.